unicode.h

```diff
@@ -0,0 +1,742 @@
+    for (auto i : other_ranges) {
+        codepoint_type_map[i] = CODEPOINT_TYPE_OTHER;
+    }
+    for (auto i : number_ranges) {
+        codepoint_type_map[i] = CODEPOINT_TYPE_NUMBER;
+    }
+    for (auto i : letter_ranges) {
+        codepoint_type_map[i] = CODEPOINT_TYPE_LETTER;
+    }
+    for (auto i : punctuation_ranges) {
+        codepoint_type_map[i] = CODEPOINT_TYPE_PUNCTUATION;
+    }
+    for (auto i : mark_ranges) {
+        codepoint_type_map[i] = CODEPOINT_TYPE_MARK;
+    }
+    for (auto i : separator_ranges) {
+        codepoint_type_map[i] = CODEPOINT_TYPE_SEPARATOR;
+    }
+    for (auto i : symbol_ranges) {
+        codepoint_type_map[i] = CODEPOINT_TYPE_SYMBOL;
+    }
+    return true;
+}
+
+static bool codepoint_type_init_search_vector() {
+    all_ranges.insert(all_ranges.end(), other_ranges.begin(), other_ranges.end());
+    all_ranges.insert(all_ranges.end(), number_ranges.begin(), number_ranges.end());
+    all_ranges.insert(all_ranges.end(), letter_ranges.begin(), letter_ranges.end());
+    all_ranges.insert(all_ranges.end(), punctuation_ranges.begin(), punctuation_ranges.end());
+    all_ranges.insert(all_ranges.end(), mark_ranges.begin(), mark_ranges.end());
+    all_ranges.insert(all_ranges.end(), separator_ranges.begin(), separator_ranges.end());
+    all_ranges.insert(all_ranges.end(), symbol_ranges.begin(), symbol_ranges.end());
+    std::sort(all_ranges.begin(), all_ranges.end());
+    return true;
+}
+
+static size_t binary_search_implement(uint32_t cp, const std::vector<std::pair<uint32_t, uint32_t>> & ranges) {
+    size_t left = 0;
+    size_t right = ranges.size() - 1;
+
+    while (left <= right) {
+        size_t mid = left + (right - left) / 2;
+        const auto& range = ranges[mid];
+
+        if (cp >= range.first && cp <= range.second) {
+            // Target is within the range of the current pair.
+            return mid;
+        } else if (cp < range.first) {
+            // Target is less than the start of the range, search in the left half.
+            right = mid - 1;
+        } else {
+            // Target is greater than the end of the range, search in the right half.
+            left = mid + 1;
+        }
+    }
+    throw std::runtime_error("Target out of range!");
+}
+
+static int codepoint_type_binary_search(uint32_t cp) {
+    try {
+        auto result = binary_search_implement(cp, all_ranges);
+        return codepoint_type_map[all_ranges[result]];
+    } catch (const std::runtime_error & e) {
+        return CODEPOINT_TYPE_UNIDENTIFIED;
+    }
+}
+
+static bool codepoint_type_init() {
+    bool map_initialized = codepoint_type_init_map();
+    bool sv_initialized = codepoint_type_init_search_vector();
+
+    if(map_initialized && sv_initialized) {
+        return true;
+    }
+
+    return false;
+}
+
+static int codepoint_type(uint32_t cp) {
+    static bool codepoint_type_initialized = codepoint_type_init();
+    return codepoint_type_binary_search(cp);
```

Comment:
The old version of doing an `unordered_map` lookup seems better. No need to implement binary search when the data structure is already available.

Reply:
Absolutely, my recent tests confirm that `unordered_map` is significantly faster, about `10X` so, but it also requires a lot more RAM, especially for full UTF-8 support. Consequently, I've opted for a hybrid approach. For high-frequency UTF-8 ranges, I'm using `unordered_map`, while for the rest, I'm employing binary search. This strategy strikes a nice balance between speed and efficiency.